crates/wasm-mutate/src/mutators/remove_item.rs

```diff
@@ -0,0 +1,1358 @@
+            ElementKind::Declared => ElementMode::Declared,
+        };
+        let element_type = self.translate_type(&element.ty)?;
+        let mut functions = Vec::new();
+        let mut exprs = Vec::new();
+        let mut reader = element.items.get_items_reader()?;
+        for _ in 0..reader.get_count() {
+            match reader.read()? {
+                ElementItem::Func(idx) => {
+                    functions.push(self.remap(Item::Function, idx)?);
+                }
+                ElementItem::Expr(expr) => match self.translate_init_expr(&expr)? {
+                    Instruction::RefFunc(n) => {
+                        exprs.push(wasm_encoder::Element::Func(n));
+                    }
+                    Instruction::RefNull(_) => {
+                        exprs.push(wasm_encoder::Element::Null);
+                    }
+                    _ => return Err(Error::no_mutations_applicable()),
+                },
+            }
+        }
+        s.segment(ElementSegment {
+            mode,
+            element_type,
+            elements: if reader.uses_exprs() {
+                Elements::Expressions(&exprs)
+            } else {
+                Elements::Functions(&functions)
+            },
+        });
+        Ok(())
+    }
+
+    fn translate_data(&mut self, data: wasmparser::Data<'_>, s: &mut DataSection) -> Result<()> {
+        let offset;
+        let mode = match &data.kind {
+            DataKind::Active {
+                memory_index,
+                init_expr,
+            } => {
+                offset = self.translate_init_expr(init_expr)?;
+                DataSegmentMode::Active {
+                    memory_index: self.remap(Item::Memory, *memory_index)?,
+                    offset: &offset,
+                }
+            }
+            DataKind::Passive => DataSegmentMode::Passive,
+        };
+        s.segment(DataSegment {
+            mode,
+            data: data.data.iter().copied(),
+        });
+        Ok(())
+    }
+
+    fn translate_code(&mut self, body: FunctionBody<'_>, s: &mut CodeSection) -> Result<()> {
+        let locals = body
+            .get_locals_reader()?
+            .into_iter()
+            .map(|local| {
+                let (cnt, ty) = local?;
+                Ok((cnt, self.translate_type(&ty)?))
+            })
+            .collect::<Result<Vec<_>>>()?;
+        let mut func = Function::new(locals);
+
+        for op in body.get_operators_reader()? {
+            let op = op?;
+            func.instruction(&self.translate_op(&op)?);
+        }
+        s.function(&func);
+        Ok(())
+    }
+
+    /// This is a pretty gnarly function that translates from `wasmparser`
+    /// operators to `wasm_encoder` operators. It's quite large because there's
+    /// quite a few wasm instructions. The theory though is that at least each
+    /// individual case is pretty self-contained.
+    fn translate_op(&mut self, op: &Operator<'_>) -> Result<Instruction<'static>> {
```

Comment:
We've implemented this kind of thing a bunch of times now. Do you want to make a `wasmparser` cargo feature for `wasm-encoder` and move all these `translate_foo` methods to be a bunch of `[Try]From<wasmparser::Foo> for Foo` implementations inside `wasm-encoder` that are gated with `cfg(feature = "wasmparser")`?

Reply:
I think this might make sense but I double-checked to ensure that at least for now this is the only translation point for `Operator`. Even if a shared translation were added this module would still have to "walk" everything to remap indices for items and such, so it's not like it'll make this module trivial. I think, though, this module could roughly look like "update a `wasmparser` thing in place" and then use shared code to translate that version to a `wasm_encoder` version. I'd prefer to make `wasmparser => wasm_encoder` infallible though so I'd want to refactor other things in `wasmparser` first like `Type` which has weird variants that are rarely used.

Reply:
Yeah, `wasmparser::Type` could definitely be cleaned up.